text/0000-template.md

```diff
@@ -0,0 +1,341 @@
+
+## Phase 3: Sync the mempool
+
+The mempool consists of transactions that have not been verified by any block in the bestchain yet. This phase can only start after phase 1 and phase 2 are finished.
+
+It is critical to have the mempool synced because (i) wallets can quickly show new transactions and update the balance, (ii) miners need the mempool to confirm transactions, and (iii) new transactions should have the most recent transactions in the mempool as parents.
+
+As the transactions form a Direct Acyclic Graph (DAG), the transactions in the mempool form a sub-DAG that connects to the sub-DAG of transactions confirmed by the bestchain (see figure below).
+
+![fig-mempool-example](./img/fig-mempool-example.svg "Mempool Example")
+
+There are several optimizations and custom rules that can be applied in this phase. For example, we can safely discard a transaction that is in conflict with another transaction in the mempool (but we can't discard both, otherwise the transaction will be lost). If it happens that the discarded transaction is confirmed, it will be downloaded as soon as the new block is synced.
+
+TODO Explain how the protocol finds what transactions need to be synced and how it downloads the transactions.
+
+XXX Is it safe to run Phase 3 in parallel with Phases 1 and 2? I feel it isn't because we don't know when it should stop.
+
+## Phase 4: Real-time propagation
+
+After the blocks and transactions are fully synced, the fullnode enters in the real-time propagation mode. It means that the fullnode will receive any new block/transaction propagated to the network, and it will propagate them to all its peers.
+
+This flood strategy is important to guarantee that all peers will receive the new blocks/transactions as soon as possible. New blocks received in this phase cannot have a parents between the genesis and the last checkpoint.
+
+## Stateful Validation
+
+As we may happen to download blocks and transactions that does not reach the genesis, we cannot guarantee we will have all parents and inputs to fully validate the blocks and transactions. So, we split the validation in the following states:
+
+1. `not-validated`: initial state, no validation has been made yet.
+
+2. `basic-validated`: all graph fields (except the parents and inputs) are valid, including the PoW, the weight field, the timestamp, etc. Funds fields have not been verified yet.
+
+3. `full-validated`: the block/transaction has been fully validated and reaches the genesis.
+
+4. `invalid`: this block/transaction is invalid and might be discarded (could be kept for quick hash detection or for banning the original peer that propagated it).
+
+Notice that `full-validated` requires the block/transaction to reach the genesis to get to this validation state. It is a critical part of this state because it guarantees that all blocks and transactions behind it are `full-validated` as well.
+
+The valid transitions between the states are shown in the figure below:
```

Comment:
An orphan block or a voided tx is 'invalid'? If yes, we can go from 'full-validated' to 'invalid' because of a reorg, no?

Reply:
Orphan blocks and voided transactions are still `full-validated`. The validation state is not related to the status (executed/voided) of the transaction.

Reply:
I added a paragraph to clarify it. Can you review it again, please?

Reply:
Great!